### PR TITLE
Adding test Capacity and App uptake features

### PR DESCRIPTION
**Pull request by sam-brett**

### CODE REVIEW

File: src/main/java/uk/co/ramp/event/AlertEventProcessor.java [hunk 1]
```diff
@@ -57,7 +61,7 @@ public ProcessedEventResult processEvent(AlertEvent event) {
     }
 
     if (nextStatus.get() != event.nextStatus()) {
-      int deltaTime = timeInStatus(nextStatus.get());
+      int deltaTime = timeInStatus(nextStatus.get(), event);
```
Comment (johnnonweiler) [inline src/main/java/uk/co/ramp/event/AlertEventProcessor.java]: The `timeInStatus` method is now doing more than calculating the time, (including e.g. recording statistics,) and should perhaps be renamed.

File: src/main/java/uk/co/ramp/event/AlertEventProcessor.java [hunk 2]
```diff
@@ -83,6 +87,7 @@ public ProcessedEventResult processEvent(AlertEvent event) {
 
     switch (event.nextStatus()) {
       case AWAITING_RESULT:
+        //        statisticsRecorder.recordTestConducted(event.time());
```
Comment (johnnonweiler) [inline src/main/java/uk/co/ramp/event/AlertEventProcessor.java]: Should this line be deleted?

File: src/main/java/uk/co/ramp/event/AlertEventProcessor.java [hunk 3]
```diff
@@ -144,4 +162,15 @@ int timeInStatus(AlertStatus newStatus) {
...
+
+  private int findNextTest(int timeNow, int capacity) {
+
+    for (int testTime = timeNow; testTime < properties.timeLimit(); testTime++) {
+      if (statisticsRecorder.getTestsConducted(testTime) < capacity) {
+        statisticsRecorder.recordTestConducted(testTime);
+        return testTime - timeNow;
+      }
+    }
+    return properties.timeLimit() * properties.timeStepsPerDay() - timeNow;
```
Comment (johnnonweiler) [inline src/main/java/uk/co/ramp/event/AlertEventProcessor.java]: In the loop above, `testTime` is set to `timeNow` and then compared to `properties.timeLimit()`, which suggests that `timeNow` and `properties.timeLimit()` have the same units.  This return statement suggests that they have different units (with one in days and the other in time steps).  Something must be wrong!

I think this is a good example of where it would be helpful to rename things to include their units ("Days" or "TimeSteps") in their names.  E.g. `properties.timeLimitDays()` or `nowTimeSteps`.  Alternatively, you could use a Time class everywhere, instead of integers, with `getDays()` and `getTimeSteps()`.

File: src/test/resources/testSeedOverride/runSettings.json [hunk 4]
```diff
@@ -2,6 +2,7 @@
   "populationSize": 10000,
   "timeLimit": 200,
   "initialExposures": 1000,
+  "seed": 3,
```
Comment (johnnonweiler) [inline src/test/resources/testSeedOverride/runSettings.json:5]: Why was this added?  (It seems strange to specify "seed" twice in the same file - it is set on line 5 and line 7 - I would have thought this should be an error.)

File: src/test/java/uk/co/ramp/people/PopulationGeneratorTest.java [hunk 5]
```diff
@@ -85,11 +84,18 @@ public void generate() {
         population.values().stream().map(Case::gender).filter(a -> a.equals(Gender.FEMALE)).count()
             / (double) popSize;
 
+    double hasApp =
+        population.values().stream().map(Case::hasApp).filter(a -> a.equals(true)).count()
```
Comment (johnnonweiler) [inline src/test/java/uk/co/ramp/people/PopulationGeneratorTest.java]: What does `filter(a -> a.equals(true))` do?  (I expect it's useful, but it's not obvious to me!)

File: src/test/java/uk/co/ramp/policy/alert/AlertContactTracerTest.java [hunk 6]
```diff
@@ -53,12 +53,29 @@ public void testPrimaryLevelTrace() {
...
+    when(population.hasApp(anyInt())).thenReturn(true);
+
     var contactTracer =
         new AlertContactTracer(tracingPolicy, eventList, population, statisticsRecorder);
 
     assertThat(contactTracer.traceRecentContacts(0, 2, 1)).containsExactlyInAnyOrder(2, 3);
   }
 
+  @Test
+  public void testPrimaryLevelTraceNoApp() {
```
Comment (johnnonweiler) [inline src/test/java/uk/co/ramp/policy/alert/AlertContactTracerTest.java:61]: This includes a lot of copy-and-pasted code from the previous test.  Could these tests be refactored to share common code?

File: input/infectionRates.json [hunk 7]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (johnnonweiler) [inline input/infectionRates.json]: This PR builds upon the changes in PR#48 - It would be easier review if _EITHER_ it did not include those changes, _OR_ the PR targetted that branch instead of the develop branch (and then I believe that if that branch is merged into develop and deleted, GitHub will automatically switch this PR to target develop).

File: src/main/java/uk/co/ramp/event/AlertEventProcessor.java [hunk 8]
```diff
@@ -115,12 +120,25 @@ public ProcessedEventResult processEvent(AlertEvent event) {
...
       case TESTED_NEGATIVE:
+        int conducted = statisticsRecorder.getTestsConducted(time);
+        int capacity =
+            (int) Math.ceil(properties.populationSize() * populationProperties.testCapacity());
+        boolean testsAvailable = conducted < capacity;
+        if (testsAvailable) {
+          statisticsRecorder.recordTestConducted(time);
+          progressionData = ImmutableMeanMax.builder().mean(1).max(1).build();
+        } else {
+          int testTime = findNextTest(time, capacity);
```
Comment (johnnonweiler) [inline src/main/java/uk/co/ramp/event/AlertEventProcessor.java:136]: I wonder if having a queue waiting until a test is available is the right way to handle running out of test capacity - I can imagine a scenario where the backlog builds up, and eventually people are waiting more than 14 days for a test.

File: src/main/java/uk/co/ramp/event/AlertEventProcessor.java [hunk 9]
```diff
@@ -115,12 +120,25 @@ public ProcessedEventResult processEvent(AlertEvent event) {
...
+          statisticsRecorder.recordTestDelayed(time, event.id());
```
Comment (johnnonweiler) [inline src/main/java/uk/co/ramp/event/AlertEventProcessor.java:137]: The statistics recorder appears to be recording things (such as event ids) which are never used.  Why is this?

File: src/main/java/uk/co/ramp/io/ContactReader.java [hunk 10]
```diff
@@ -45,6 +45,7 @@ public ContactReader(
 
   List<ImmutableContactEvent> resampleContacts(List<ImmutableContactEvent> contactEvents) {
     int steps = properties.timeStepsPerDay();
+    int offset = properties.dayOffset().orElse(0);
```
Comment (johnnonweiler) [inline src/main/java/uk/co/ramp/io/ContactReader.java:48]: I don't know if you're generally using GitHub's "squash commits and merge" option, but it is probably a good idea here.  It would be good to mention the addition of the "dayOffset" parameter in the new commit message, as well as explaining the capacity and app uptake changes.

Also, I don't know why `input/contacts.csv` was changed in this PR, but that should be explained too.

**Review by johnnonweiler — DISMISSED** [review]
This PR includes a number of changes, some which are unrelated to each other, and that makes it a bit difficult to review.  I think it's all ok.

File: src/main/java/uk/co/ramp/event/AlertEventProcessor.java (hunk 3 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/event/AlertEventProcessor.java]: Good spot. I'll change timeLimit to timeLimitDays

File: src/test/resources/testSeedOverride/runSettings.json (hunk 4 above)
Comment (sam-brett) [reply src/test/resources/testSeedOverride/runSettings.json:5]: I think this might have been a merge problem.

File: src/test/java/uk/co/ramp/people/PopulationGeneratorTest.java (hunk 5 above)
Comment (sam-brett) [reply src/test/java/uk/co/ramp/people/PopulationGeneratorTest.java]: replaced with ```population.values().stream().filter(Case::hasApp).count()```

File: src/test/java/uk/co/ramp/policy/alert/AlertContactTracerTest.java (hunk 6 above)
Comment (sam-brett) [reply src/test/java/uk/co/ramp/policy/alert/AlertContactTracerTest.java:61]: refactored

File: src/main/java/uk/co/ramp/event/AlertEventProcessor.java (hunk 8 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/event/AlertEventProcessor.java:136]: This is just a first take, Sibylle and Louise seem happy with the simple delay, but we can discuss improvements further.

File: src/main/java/uk/co/ramp/event/AlertEventProcessor.java (hunk 9 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/event/AlertEventProcessor.java:137]: this is a hook in case they want to output information beyond the total number of delays.

File: src/main/java/uk/co/ramp/io/ContactReader.java (hunk 10 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/io/ContactReader.java:48]: contact.csv didn't have the label field, so I added it.

File: src/test/resources/testSeedOverride/runSettings.json (hunk 4 above)
Comment (johnnonweiler) [reply src/test/resources/testSeedOverride/runSettings.json:5]: I fear you've had another merge problem.

**Review by johnnonweiler — APPROVED** [review]
